### DISCUSSION TITLE
Latest Breaking Version 3.x

### DISCUSSION THREAD

**Discussion by ghiscoding** [ghiscoding/aurelia-slickgrid#485] [General]
Latest Update on the upcoming breaking version 3.x
Anyone having major issues with the new release can report in this thread.

Things left to do to be fully functional and parity with previous version 2.x
- [x] Collection Watch for AutoComplete Filter & SingleSelect, MultipleSelect Filters 
   - it's in the work ([PR #208](https://github.com/ghiscoding/slickgrid-universal/pull/208)), nearly done with the feature, just need few more tests (unit tests + E2E tests)
- [x] RequireJS still throws some errors unless `@slickgrid-universal/common` is added to the `aurelia.json` dependencies mapping
   - I know what is causing it, it was identified [here](https://github.com/ghiscoding/aurelia-slickgrid-demos/pull/5#issuecomment-748569748) with the help of `3cp` which was super helpful.

**Comment by ghiscoding** [comment]
@zewa666 I pushed another version, wanted to push a Beta version but somehow I misconfigured the command to use with `standard-version`, it was supposed to be a prerelease but that command didn't work so anyway, long story short `3.0.0` is somehow released (even though I'm not totally finished). Give it a try, I've merged the Aurelia-Slickgrid-Demos as well with latest code (note the only difference from the Alpha to now was the grid option `registerExternalServices` is now `registerExternalResources` for both Export Services & Components (Migration was updated as well).

P.S. I'm trying out this new Discussion thing, it seems similar to how issues work just without it being an issue hehe

**Reply by zewa666** [reply]
Yeah perfect wanted to propose this already

**Comment by ghiscoding** [comment]
@zewa666 
So I believe that I fixed everything now, that includes the following
- fixed RequireJS auto-register dependencies (you can now remove all the Slickgrid-Universal mappings, see this demos [PR](https://github.com/ghiscoding/aurelia-slickgrid-demos/pull/6/files))
   - `3cp` helped me identified 2 problems (the invalid `browser` path in `package.json` and also some use of `require` to load SlickGrid plugins dynamically that were actually not loaded at all in RequireJS).  
- added the Collection Watch (see [Wiki](https://github.com/ghiscoding/aurelia-slickgrid/wiki/Select-Filter#collection-watch)) that was missing from Aurelia-Slickgrid version 2.x (it is now directly in Slickgrid-Universal)

So I think I'm finally done with all of it, let me know if you give it a try, this year or the next 😉

**Reply by zewa666** [reply]
This is fantastic. Thanks for the efforts and I'll ping you once checked out. As far as I understand now work mainly continues in universal and merely examples plus version bumps happen in this repo right?

**Reply by ghiscoding** [reply]
Yeah you pretty much sums it all up. 
I can finally now do fixes in 1 place and expect them to bubble up to Aurelia-Slickgrid, which was the plan all along 🚀

**Comment by ghiscoding** [comment]
Well actually now that I think of it, there's actually the subject of proper Dependency Injection but even the manual injection works just fine. The main struggle is to find something that would work on both side, that would require some kind of DI that is framework agnostic. I was close to have something working with some code I found over the net but it starts to fail with 2 things (abstract classes and optional services), I use abstract classes mainly for the pub/sub & translate service which the latter can also be optional... At the end, I might keep the manual DI because it works and it has no cost in size while going for other agnostic DI libs might increase the lib size for no real benefit, except for internal cleaner and more readable code. [TypeDI](https://docs.typestack.community/typedi/v/develop/01-getting-started) might be 1 of those agnostic DI that could potentially work, will give it a try eventually, just worry about the size that it adds

**Comment by ArnaudVanHerck** [comment]
The latest version 3.5.0 gives us the following error `export 'BackendUtilityService' was not found in '@slickgrid-universal/common'`. Somebody else experiencing this?

**Reply by ghiscoding** [reply]
that was indeed refactored code for Backend Services in last version of Slickgrid-Universal, my guess is that you didn't update all Slickgrid-Universal packages, so make sure that **all** Slickgrid-Universal packages that you use are at version `0.12.0`. Every version release of Aurelia-Slickgrid goes hand in hand with Slickgrid-Universal, you need to update both side at same time.

**Reply by ArnaudVanHerck** [reply]
The only one we're using is excel export which we have at 0.12.0
But I guess it could be npm not updating node modules correctly 😕 I'll try a bit more...

**Reply by ghiscoding** [reply]
FYI, you might be thinking that just using the `^` will update all Slickgrid-Universal but it won't and 3cp told me that in another package. I didn't know that `0.x.x` have special rule. I might just bump all Slickgrid-Universal packages to `1.x.x` in the near future, probably after I get Angular-Slickgrid to use it as well.

![image](https://user-images.githubusercontent.com/643976/112473724-1afd4d80-8d45-11eb-8dfb-54c33c4c0a9c.png)

**Reply by ghiscoding** [reply]
another thing you could try is to force the download, though I didn't have to do that last night when I updated the Aurelia-Slickgrid-Demos, is to delete your lock file and the 2 folders in your `node_modules\@slickgrid-universal` and `\aurelia-slickgrid` to make sure they download the latest `0.12.0` everywhere. Search for Slickgrid-Universal in your lock file, it should tell you easily what got updated or not

**Reply by ArnaudVanHerck** [reply]
I got it to work, it was like you said, an issue with npm versions and node_modules not updating correctly. My dev PC doesn't like me I think 😄

**Reply by ghiscoding** [reply]
nice to hear it works, now you can take advantage of the PR you contributed 😉

**Reply by ArnaudVanHerck** [reply]
Yes, exactly. Thank you for all your hard work in this repo! 😃
